File: storage/ndb/src/kernel/blocks/ndbfs/Filename.cpp
/*
   Copyright (c) 2003, 2023, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include <ndb_global.h>

#include "Filename.hpp"
#include "ErrorHandlingMacros.hpp"
#include "RefConvert.hpp"
#include "DebuggerNames.hpp"
#include "Ndbfs.hpp"

#include <signaldata/FsOpenReq.hpp>

#define JAM_FILE_ID 383


static const char* fileExtension[] = {
  ".Data",
  ".FragLog",
  ".LocLog",
  ".FragList",
  ".TableList",
  ".SchemaLog",
  ".sysfile",
  ".log",
  ".ctl"
};

static const Uint32 noOfExtensions = sizeof(fileExtension)/sizeof(char*);

Filename::Filename()
{
}

Filename::~Filename(){
}

void 
Filename::set(Ndbfs* fs,
              BlockReference blockReference,
              const Uint32 filenumber[4], bool dir,
              SegmentedSectionPtr ptr)
{
  m_base_path_spec = FsOpenReq::BP_MAX;

  char buf[PATH_MAX];

  const Uint32 type = FsOpenReq::getSuffix(filenumber);
  const Uint32 version = FsOpenReq::getVersion(filenumber);

  size_t sz;
  if (version == 2)
  {
    m_base_path_spec = FsOpenReq::BP_BACKUP;
    sz = BaseString::snprintf(theName, sizeof(theName), "%s", 
                              fs->get_base_path(FsOpenReq::BP_BACKUP).c_str());
    m_base_name = theName + fs->get_base_path(FsOpenReq::BP_BACKUP).length();
  }
  else
  {
    m_base_path_spec = FsOpenReq::BP_FS;
    sz = BaseString::snprintf(theName, sizeof(theName), "%s", 
                              fs->get_base_path(FsOpenReq::BP_FS).c_str());
    m_base_name = theName + fs->get_base_path(FsOpenReq::BP_FS).length();
  }
  
  switch(version){
  case 1 :{
    const Uint32 diskNo = FsOpenReq::v1_getDisk(filenumber);
    const Uint32 table  = FsOpenReq::v1_getTable(filenumber);
    const Uint32 frag   = FsOpenReq::v1_getFragment(filenumber);
    const Uint32 S_val  = FsOpenReq::v1_getS(filenumber);
    const Uint32 P_val  = FsOpenReq::v1_getP(filenumber);

    if (diskNo < 0xff){	  
      BaseString::snprintf(buf, sizeof(buf), "D%d%s", diskNo, DIR_SEPARATOR);
      strcat(theName, buf);
    }
    
    {
      const char* blockName = getBlockName( refToMain(blockReference) );
      if (blockName == NULL){
	ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"","No Block Name");
	return;
      }
      BaseString::snprintf(buf, sizeof(buf), "%s%s", blockName, DIR_SEPARATOR);
      strcat(theName, buf);
    }
    
    if (table < 0xffffffff){
      BaseString::snprintf(buf, sizeof(buf), "T%d%s", table, DIR_SEPARATOR);
      strcat(theName, buf);
    }
    
    if (frag < 0xffffffff){
      BaseString::snprintf(buf, sizeof(buf), "F%d%s", frag, DIR_SEPARATOR);
      strcat(theName, buf);
    }
    
    
    if (S_val < 0xffffffff){
      BaseString::snprintf(buf, sizeof(buf), "S%d", S_val);
      strcat(theName, buf);
    }

    if (P_val < 0xff){
      BaseString::snprintf(buf, sizeof(buf), "P%d", P_val);
      strcat(theName, buf);
    }
    
  }
  break;
  case 2:{
    const Uint32 seq = FsOpenReq::v2_getSequence(filenumber);
    const Uint32 nodeId = FsOpenReq::v2_getNodeId(filenumber);
    const Uint32 partNum =  FsOpenReq::v2_getPartNum(filenumber);
    const Uint32 totalParts =  FsOpenReq::v2_getTotalParts(filenumber);
    const Uint32 count = FsOpenReq::v2_getCount(filenumber);

    if(partNum == 0)
      BaseString::snprintf(buf, sizeof(buf), "BACKUP%sBACKUP-%u%s",
            DIR_SEPARATOR, seq, DIR_SEPARATOR);
    else
      BaseString::snprintf(buf, sizeof(buf), "BACKUP%sBACKUP-%u%sBACKUP-%u-PART-%u-OF-%u%s",
             DIR_SEPARATOR, seq, DIR_SEPARATOR, seq, partNum, totalParts, DIR_SEPARATOR);

    strcat(theName, buf);
    if(count == 0xffff) {
      BaseString::snprintf(buf, sizeof(buf), "BACKUP-%u.%u",
	       seq, nodeId); strcat(theName, buf);
    } else {
      BaseString::snprintf(buf, sizeof(buf), "BACKUP-%u-%u.%u",
	       seq, count, nodeId); strcat(theName, buf);
    }
    break;
  }
  break;
  case 3:{
    const Uint32 diskNo = FsOpenReq::v1_getDisk(filenumber);

    if(diskNo == 0xFF){
      ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"","Invalid disk specification");
    }

    BaseString::snprintf(buf, sizeof(buf), "D%d%s", diskNo, DIR_SEPARATOR);
    strcat(theName, buf);
  }
    break;
  case 4:
  {
    const unsigned ptr_sz_bytes = ptr.sz * 4;
    if (ptr_sz_bytes > PATH_MAX) {
      ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"",
                "File name is too long");
      return;
    }
    if (ptr_sz_bytes == 0) {
      ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"",
                "File name is empty");
      return;
    }

    char buf[PATH_MAX];
    copy((Uint32*)&buf[0], ptr);

    const bool nul_terminated = ((buf[ptr_sz_bytes-1] == '\0') ||
                                 (buf[ptr_sz_bytes-2] == '\0') ||
                                 (buf[ptr_sz_bytes-3] == '\0') ||
                                 (buf[ptr_sz_bytes-4] == '\0'));
    if (!nul_terminated) {
      ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"",
                "File name is not NUL-terminated");
      return;
    }
    if(buf[0] == '\0')
    {
      ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"",
                "File name is not given");
      return;
    }

    const unsigned theName_sz = sizeof(theName);
    if(buf[0] == DIR_SEPARATOR[0])
    {
      BaseString::snprintf(theName, theName_sz, "%s", buf);
      m_base_name = theName;
    }
    else
    {
#ifdef _WIN32
      char* b= buf;
      while((b= strchr(b, '/')) && b)
      {
        *b= '\\';
      }
#endif
      Uint32 bp = FsOpenReq::v4_getBasePath(filenumber);
      m_base_path_spec = bp;
      const Uint32 base_path_len = fs->get_base_path(bp).length();
      const size_t concat_sz = BaseString::snprintf(theName, theName_sz, "%s%s",
                                fs->get_base_path(bp).c_str(), buf);
      if (concat_sz >= theName_sz) {
        // File path name is truncated
        ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"",
                  "File path name is too long");
        return;
      }
      m_base_name = theName + base_path_len;
    }
    return; // No extension
  }
  case 5:
  {
    Uint32 tableId = FsOpenReq::v5_getTableId(filenumber);
    Uint32 lcpNo = FsOpenReq::v5_getLcpNo(filenumber);
    Uint32 fragId = FsOpenReq::v5_getFragmentId(filenumber);
    BaseString::snprintf(buf, sizeof(buf), "LCP%s%d%sT%dF%d", DIR_SEPARATOR, lcpNo, DIR_SEPARATOR, tableId, fragId);
    strcat(theName, buf);
    break;
  }
  case 6:
  {
    Uint32 bp = FsOpenReq::v5_getLcpNo(filenumber);
    m_base_path_spec = bp;
    sz = BaseString::snprintf(theName, sizeof(theName), "%s",
                              fs->get_base_path(bp).c_str());
    break;
  }
  default:
    ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"","Wrong version");
  }
  if (type >= noOfExtensions){
    ERROR_SET(ecError, NDBD_EXIT_AFS_PARAMETER,"","File Type doesn't exist");
    return;
  }
  strcat(theName, fileExtension[type]);
  
  if(dir == true){
    for(int l = (int)strlen(theName) - 1; l >= 0; l--){
      if(theName[l] == DIR_SEPARATOR[0]){
	theName[l] = 0;
	break;
      }
    }
  }
}

File: libs/scripts/impl/file_name.py
#!/usr/bin/python3.11
# Copyright (c) 2024, Oracle and/or its affiliates.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2.0,
# as published by the Free Software Foundation.
#
# This program is designed to work with certain software (including
# but not limited to OpenSSL) that is licensed under separate terms,
# as designated in a particular file or component or in included license
# documentation.  The authors of MySQL hereby grant you an additional
# permission to link the program and your derivative works with the
# separately licensed software that they have either included with
# the program or referenced in the documentation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License, version 2.0, for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA

'''
Helper functions used in move/create scripts, for working with filenames.
'''

from   messages import log_error
import string

import os
from   typing import List, Tuple

def split_existing_nonexisting_path(path: str) -> Tuple[str, str]:
    '''Split path into prefix of existing and suffix of nonexisting directories.

    Return a 2-tuple where the first component is the prefix of existing
    directories and the suffix is the suffix of non-existing directories.
    '''
    suffix = ''
    prefix = path
    while not os.path.exists(prefix):
        prefix, part = os.path.split(prefix)
        if suffix:
            suffix = os.path.join(part, suffix)
        else:
            suffix = part
    return prefix, suffix

def validate_path(path: str) -> None:
    '''Check that the path has a valid format, and report an error otherwise.'''

    if not path.startswith('libs/mysql'):
        log_error(
            f'Path "{path}" does not begin with libs/mysql.'
        )
    if '//' in path:
        log_error(
            f'Path {path} contains double slashes.'
        )
    if path.startswith('//'):
        log_error(
            f'Path {path} starts with slash.'
        )
    if path.endswith('//'):
        log_error(
            f'Path {path} ends with slash.'
        )
    _, nonexisting = split_existing_nonexisting_path(path)
    if not nonexisting: return
    if not string.re_fullmatch(
        pattern=r'(/[a-z][a-z0-9]*(_[a-z0-9]+)*)*(\.cc|\.h)?',
        string=f'/{nonexisting}',
    ):
        log_error(
            f'Path "{path}" is not in the correct format. '
            f'The part of it that does not already exist is "{nonexisting}", '
            'which has to follow the following naming rules. '
            'Each path component may consist of one or more words, separated '
            'by underscores. '
            'Each word must contain one or more letters, all from the set '
            '[a-z0-9]. '
            'The first letter of the first word in each path component must '
            'be in the set [a-z].'
        )

def normalize_path(path: str) -> str:
    '''Add the prefix libs/mysql if needed.

    This assumes that paths that don't start with ./ refer to libs/mysql, and
    add that to the path unless it is there already.
    '''

    path = path.removeprefix('./libs/mysql/')
    path = path.removeprefix('libs/mysql/')
    path = path.removeprefix('mysql/')
    if path[0] in '~./':
        log_error(f'Can only create files in libs/mysql, not in {path}')
    path = 'libs/mysql/' + path
    return path

# File type constants
class Source: pass
class Header: pass
class Markdown: pass
class Directory: pass
class Cmakelists: pass
FileType = (
    type[Source] | type[Header] | type[Markdown] | type[Directory] |
    type[Cmakelists]
)

cmakelists_name = 'CMakeLists.txt'
libs_dox_path = 'libs/mysql/libs.dox'

def file_type_by_extension(path: str) -> FileType:
    '''Determine the file type by the extension.

    Parameters
    ----------
    :param path: The path to check.
    :returns: One of the objects Header, Source, Markdown, or Directory.
    '''

    ext = os.path.splitext(path)[1]
    if ext in ('.h', '.hh', '.hpp', '.hxx'):
        return Header
    if ext in ('.c', '.cc', '.cpp', '.cxx'):
        return Source
    if ext == '.md':
        return Markdown
    if ext == '':
        return Directory
    if os.path.basename(path) == cmakelists_name:
        return Cmakelists
    return ext

def source_to_headers(path: str) -> List[str]:
    '''Returns the list of header(s) for a given file.

    This returns the 8 filenames with extensions [-impl]{.h,.hh,.hpp,.hxx}, and
    the first one is guaranteed to be the good old .h file.
    '''

    file_type = file_type_by_extension(path)
    if file_type == Source:
        noext = os.path.splitext(path)[0]
        return [
            f'{noext}{infix}{extension}'
            for infix in ('', '-impl')
            for extension in ('.h', '.hh', '.hpp', '.hxx')
        ]
    raise ValueError(f"Not a source file: {path}")
